pkg/providers/vsphere/vmprovider_vm.go

```diff
@@ -1486,16 +1488,188 @@ func (vs *vSphereVMProvider) getVMYamlFromSnapshot(
+			ImageName: "",
+			Image:     nil,
+
+			Cdrom: []vmopv1.VirtualMachineCdromSpec{},
+
+			// We only support VM snapshots and affinity policies are not part of VMs.
+			Affinity: nil,
+
+			// It's not possible to construct the crypto spec looking at the VM.
+			Crypto: nil,
+
+			// It doesn't matter since in VCD, VMs can't be customized upon subsequent boots.  Except using Force Customization which is an admin operation.
+			Bootstrap: nil,
+
+			// These are set to the field defaults
+			PowerOffMode:     vmopv1.VirtualMachinePowerOpModeTrySoft,
+			SuspendMode:      vmopv1.VirtualMachinePowerOpModeTrySoft,
+			RestartMode:      vmopv1.VirtualMachinePowerOpModeTrySoft,
+			PromoteDisksMode: vmopv1.VirtualMachinePromoteDisksModeOnline,
+
+			// No PVCs for imported VMs.  For day 2, users can convert disks in PVCs using just-in-time PVCs.
+			Volumes: []vmopv1.VirtualMachineVolume{},
+
+			// This is a VM service specific property, so it does not make sense for VCD imported VM.
+			ReadinessProbe: nil,
+
+			// This is used for slot reservation starting VCF 9.0.  The reverted VM will continue to use the reserved
+			// profile but it won't be part of the spec since the plumbing from VM class to VCFA won't exist (VCFA
+			// expects slots of a class to be reserved if this is set).
+			Reserved: nil,
+
+			// The VM will continue to have the same boot options from the snapshot state.
+			// If user wants to change that, they can specify a custom boot option (boot order).
+			BootOptions: nil,
+
+			// currentSnapshot is only used to trigger a revert, so this must be empty.
+			CurrentSnapshot: nil,
+
+			// Group name is overridden to the VM's current group.
+			// We can't know (or guess) what the group the VM was when the snapshot was taken.
+			GroupName: vmCtx.VM.Spec.GroupName,
+
+			Advanced: &vmopv1.VirtualMachineAdvancedSpec{
+				ChangeBlockTracking: moSnap.Config.ChangeTrackingEnabled,
+			},
+
+			Network: &vmopv1.VirtualMachineNetworkSpec{
+				HostName: vmCtx.MoVM.Guest.HostName,
+			},
+		}
+
+		pbmClient, err := pbm.NewClient(vmCtx, vcVM.Client())
+		if err != nil {
+			vmCtx.Logger.Error(err, "error creating pbm client")
+			return "", err
+		}
+		profileIDs, err := pbmClient.QueryAssociatedProfile(vmCtx, pbmtypes.PbmServerObjectRef{
+			ObjectType: "virtualMachine",
+			Key:        vmCtx.VM.Status.UniqueID,
+			ServerUuid: moSnap.Config.InstanceUuid,
+		})
+		if err != nil {
+			vmCtx.Logger.Error(err, "error querying pbm client")
+			return "", err
+		}
+
+		if len(profileIDs) == 0 {
+			vmCtx.Logger.V(4).Info("VM is not associated with any profile")
+		} else {
+			// Taking only the first one for now
+			// TODO(nabarun): Once consensus is established on what would be the storage class when multiple storage
+			// policies are returned, update this logic if needed
+			profileID := profileIDs[0]
+			storageClasses, err := spq.GetStorageClassesForPolicy(vmCtx, vs.k8sClient, vmCtx.VM.Namespace, profileID.UniqueId)
```

Comment:
It turns out that when you take a snapshot of a VM using storage policy `A`, and then update / modify the VM to use policy `B`, and then revert to the snapshot, the VM continues to have the policy `A`.  So, let's just simplify all of this and set the VM's spec.storageClass to the one from the current VM (like we do for groupName).